// @opensumi/ide-utils/lib/argv.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArgvFactory = exports.parseString = exports.parseArgv = void 0;
const tslib_1 = require("tslib");
/* eslint-disable no-console */
const mri_1 = tslib_1.__importDefault(require("mri"));
function parseArgv(argv, options) {
    return (0, mri_1.default)(argv, options);
}
exports.parseArgv = parseArgv;
function parseString(str, options) {
    return (0, mri_1.default)(str.split(' '), options);
}
exports.parseString = parseString;
class ArgvFactory {
    constructor(factoryArgv) {
        this.factoryArgv = factoryArgv;
        this._string = [];
    }
    get argv() {
        var _a;
        if (!this._argv) {
            const options = {
                string: this._string,
            };
            this._argv = (0, mri_1.default)(this.factoryArgv, options);
        }
        if (((_a = this._argv) === null || _a === void 0 ? void 0 : _a.help) && this.showHelp) {
            this.showUsage();
            process.exit(0);
        }
        return this._argv;
    }
    usage(message) {
        this._usage = message;
        return this;
    }
    showUsage() {
        if (this._usage) {
            console.log(this._usage.trimLeft());
            if (this.showHelp) {
                console.log('Usage:');
                console.log('    ' + this.showHelp);
            }
        }
    }
    string(k) {
        this._string.push(k);
        return this;
    }
    help() {
        this.showHelp = '--help Show help';
        return this;
    }
}
exports.ArgvFactory = ArgvFactory;
