// @slardar/sdk-web/esm/plugins/monitors/blank-screen.js
import { __read } from "tslib";
import { BlankScreenMonitor } from '../../monitors/blank-screen';
import { DEFAULT_THRESHOLD, DEFAULT_T0_DATA_URL_QUALITY } from '../../monitors/blank-screen/util';
import { applyMonitor, getConfig } from '../../utils';
import { getPluginConfig } from './utils';
export var BLANK_SCREEN_MONITOR_PLUGIN_NAME = 'blankScreen';
function BlankScreenMonitorBasePlugin(client, defaultConfig, overrideConfig) {
    client.on('init', function () {
        var config = getPluginConfig(client, BLANK_SCREEN_MONITOR_PLUGIN_NAME, defaultConfig) ||
            getConfig(overrideConfig, defaultConfig);
        if (!config) {
            return;
        }
        var _a = __read(applyMonitor(BlankScreenMonitor, config, client.report.bind(client)), 3), tearDown = _a[0], applyOtherReport = _a[1], detectBlankScreen = _a[2];
        var listenOtherReport = function (ev) {
            applyOtherReport(ev);
            return ev;
        };
        client.on('report', listenOtherReport);
        client.on('beforeDestroy', function () {
            tearDown();
            client.off('report', listenOtherReport);
        });
        client.provide('detectBlankScreen', detectBlankScreen);
    });
}
var defaultByteConfig = {
    autoDetect: true,
    threshold: DEFAULT_THRESHOLD,
    screenshot: true,
    ssUrl: 'https://sf16-short-va.bytedapm.com/slardar/fe/sdk-web/screenshot.min.js',
    mask: false,
    partialShot: false,
    quality: DEFAULT_T0_DATA_URL_QUALITY,
};
export function BlankScreenMonitorPlugin(client, overrideConfig) {
    BlankScreenMonitorBasePlugin(client, defaultByteConfig, overrideConfig);
}
var defaultVolConfig = {
    autoDetect: true,
    threshold: DEFAULT_THRESHOLD,
    screenshot: true,
    ssUrl: 'https://apm.volccdn.com/mars-web/apmplus/web/html2canvas.min.js',
    mask: false,
    partialShot: false,
    quality: DEFAULT_T0_DATA_URL_QUALITY,
};
export function BlankScreenMonitorVolPlugin(client, overrideConfig) {
    BlankScreenMonitorBasePlugin(client, defaultVolConfig, overrideConfig);
}
