// @lark-opdev/cli/libs/utils/enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GadgetPlatform = exports.Host = exports.CacheType = exports.mode = exports.platform = exports.TypeEnum = void 0;
const upload_1 = require("@bdeefe/feishu-devtools-core/libs/types/upload");
Object.defineProperty(exports, "GadgetPlatform", { enumerable: true, get: function () { return upload_1.GadgetPlatform; } });
var commonData_1 = require("@bdeefe/feishu-devtools-core/libs/types/commonData");
Object.defineProperty(exports, "TypeEnum", { enumerable: true, get: function () { return commonData_1.AppTypeEnum; } });
exports.platform = Object.values(upload_1.GadgetPlatform);
exports.mode = Object.values(upload_1.GadgetPCMode);
var CacheType;
(function (CacheType) {
    CacheType["update"] = "update";
    CacheType["jssdk"] = "jssdk";
    CacheType["all"] = "all";
})(CacheType = exports.CacheType || (exports.CacheType = {}));
var Host;
(function (Host) {
    Host["docs"] = "docs";
    Host["workplace"] = "workplace";
})(Host = exports.Host || (exports.Host = {}));
