// @slardar/sdk-web/esm/collector/client.js
import { noop } from '@slardar/sdk-template';
import { getLocationUrl } from '../utils/url';
var getOverrides = function (config) {
    return {
        pid: config.pid,
        view_id: config.viewId,
        url: getLocationUrl(),
    };
};
export var PV_SUBJECT_NAME = 'view_0';
/**
 * 由于跟当前实例强相关 所以不能挂在global上
 */
export var applyPV = function (client) {
    return function (next, tearDown) {
        var currentOverrides = getOverrides(client.config());
        var handleConfig = function (newConfig) {
            var _a;
            if (newConfig.viewId && newConfig.viewId !== ((_a = client.config()) === null || _a === void 0 ? void 0 : _a.viewId)) {
                next(currentOverrides);
                currentOverrides = getOverrides(newConfig);
            }
        };
        client.on('beforeConfig', handleConfig);
        tearDown(function () {
            client.off('beforeConfig', handleConfig);
        });
    };
};
export var FIRST_PV_SUBJECT_NAME = 'f_view_0';
/**
 * 由于跟当前实例强相关 所以不能挂在global上
 */
export var applyFirstPV = function (client) {
    return function (_, tearDown) {
        var firstOverrides = getOverrides(client.config());
        tearDown(noop, function (subscriber) {
            firstOverrides && subscriber(firstOverrides);
        });
    };
};
