// @slardar/electron/esm/plugins/__test__/utils/getMockData.js
import sinon from 'sinon';
export var getMockData = function () {
    var collectCbs = {
        asyncSample: sinon.stub().callsFake(function () { return new Promise(function (res) { return setTimeout(res, 1000); }); }),
        syncSample: sinon.stub(),
    };
    var reportCbs = {
        asyncSample: sinon.stub(),
        syncSample: sinon.stub(),
    };
    var transSucCbs = {
        asyncSample: sinon.stub(),
        syncSample: sinon.stub(),
    };
    var sampleConfigList = [
        {
            name: 'asyncSample',
            collectInterval: 200,
            sampleRate: 1,
            collectCb: collectCbs['asyncSample'],
            reportCb: reportCbs['asyncSample'],
            transSucCb: transSucCbs['asyncSample'],
            async: {
                isWaitTillReturn: true,
            },
        },
        {
            name: 'syncSample',
            collectInterval: 200,
            sampleRate: 1,
            collectCb: collectCbs['syncSample'],
            reportCb: reportCbs['syncSample'],
            transSucCb: transSucCbs['syncSample'],
        },
    ];
    return { sampleConfigList: sampleConfigList, collectCbs: collectCbs, reportCbs: reportCbs, transSucCbs: transSucCbs };
};
