// @lark-opdev/cli/libs/implements/upload.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const feishu_devtools_core_1 = require("@bdeefe/feishu-devtools-core");
const base_1 = __importDefault(require("./base"));
class Upload extends base_1.default {
    constructor() {
        super(feishu_devtools_core_1.ModuleName.upload);
    }
    gadget(options) {
        return this.manager.execute(feishu_devtools_core_1.ModuleName.upload, feishu_devtools_core_1.UploadCmdEnum.uploadGadget, {
            compilePath: options.path,
            platform: options.platform,
            version: options.version,
            description: options.description,
            useNPM: options.useNPM,
        });
    }
    gadgetPlugin(options) {
        return this.manager.execute(feishu_devtools_core_1.ModuleName.upload, feishu_devtools_core_1.UploadCmdEnum.uploadGadgetPlugin, {
            compilePath: options.path,
            version: options.version,
            description: options.description,
        });
    }
    block(options) {
        return this.manager.execute(feishu_devtools_core_1.ModuleName.upload, feishu_devtools_core_1.UploadCmdEnum.uploadBlock, {
            compilePath: options.path,
            version: options.version,
            description: options.description,
        });
    }
}
exports.default = Upload;
