// @opensumi/ide-core-common/lib/types/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtensionDidContributes = exports.ExtensionActivateEvent = exports.ExtensionEnabledEvent = exports.asExtensionCandidate = exports.ExtensionConnectModeOption = exports.ExtensionCandidate = void 0;
const event_bus_1 = require("../event-bus");
class ExtensionCandidate {
}
exports.ExtensionCandidate = ExtensionCandidate;
var ExtensionConnectModeOption;
(function (ExtensionConnectModeOption) {
    ExtensionConnectModeOption[ExtensionConnectModeOption["TCP"] = 0] = "TCP";
    ExtensionConnectModeOption[ExtensionConnectModeOption["IPC"] = 1] = "IPC";
})(ExtensionConnectModeOption = exports.ExtensionConnectModeOption || (exports.ExtensionConnectModeOption = {}));
/**
 * 将插件路径转换为 ExtensionCandidate 对象
 * @param extensionPath 插件路径
 * @param isDevelopment 是否为开发模式下加载的插件
 */
function asExtensionCandidate(extensionPath, isDevelopment = false) {
    return { path: extensionPath, isBuiltin: true, isDevelopment };
}
exports.asExtensionCandidate = asExtensionCandidate;
class ExtensionEnabledEvent extends event_bus_1.BasicEvent {
}
exports.ExtensionEnabledEvent = ExtensionEnabledEvent;
class ExtensionActivateEvent extends event_bus_1.BasicEvent {
}
exports.ExtensionActivateEvent = ExtensionActivateEvent;
class ExtensionDidContributes extends event_bus_1.BasicEvent {
}
exports.ExtensionDidContributes = ExtensionDidContributes;
