// @opensumi/ide-core-common/lib/types/markers/markers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MarkerSeverity = exports.MapMap = void 0;
const ide_utils_1 = require("@opensumi/ide-utils");
var MapMap;
(function (MapMap) {
    function get(map, key1, key2) {
        if (map[key1]) {
            return map[key1][key2];
        }
        return undefined;
    }
    MapMap.get = get;
    function getMap(map, key) {
        return map[key];
    }
    MapMap.getMap = getMap;
    function set(map, key1, key2, value) {
        if (!map[key1]) {
            map[key1] = Object.create(null);
        }
        map[key1][key2] = value;
    }
    MapMap.set = set;
    function setMap(map, key, value) {
        map[key] = value;
    }
    MapMap.setMap = setMap;
    function remove(map, key1, key2) {
        if (map[key1] && map[key1][key2]) {
            delete map[key1][key2];
            if ((0, ide_utils_1.isEmptyObject)(map[key1])) {
                delete map[key1];
            }
            return true;
        }
        return false;
    }
    MapMap.remove = remove;
    function removeMap(map, key) {
        if (map[key]) {
            const result = map[key];
            delete map[key];
            return result;
        }
        return undefined;
    }
    MapMap.removeMap = removeMap;
    function keys(map) {
        const result = [];
        if (map) {
            // eslint-disable-next-line guard-for-in
            for (const key in map) {
                result.push(key);
            }
        }
        return result;
    }
    MapMap.keys = keys;
})(MapMap = exports.MapMap || (exports.MapMap = {}));
/**
 * Marker的严重程度
 */
var MarkerSeverity;
(function (MarkerSeverity) {
    MarkerSeverity[MarkerSeverity["Hint"] = 1] = "Hint";
    MarkerSeverity[MarkerSeverity["Info"] = 2] = "Info";
    MarkerSeverity[MarkerSeverity["Warning"] = 4] = "Warning";
    MarkerSeverity[MarkerSeverity["Error"] = 8] = "Error";
})(MarkerSeverity = exports.MarkerSeverity || (exports.MarkerSeverity = {}));
(function (MarkerSeverity) {
    function compare(a, b) {
        return b - a;
    }
    MarkerSeverity.compare = compare;
})(MarkerSeverity = exports.MarkerSeverity || (exports.MarkerSeverity = {}));
