// @slardar/sdk-web/esm/collector/__test__/mock.js
/* eslint-disable @typescript-eslint/prefer-nullish-coalescing */
export function mockHistoryAndLocation(initHref) {
    var stack = [
        {
            state: {},
            title: '',
            url: initHref,
        },
    ];
    var location = {
        get href() {
            return stack.length ? stack[stack.length - 1].url : 'about:blank';
        },
    };
    var history = {
        pushState: function (state, title, url) {
            stack.push({
                state: state,
                title: title,
                url: url || location.href,
            });
        },
        replaceState: function (state, title, url) {
            var s = {
                state: state,
                title: title,
                url: url || location.href,
            };
            stack.pop();
            stack.push(s);
        },
    };
    var setHref = function (url) {
        stack.push({
            state: {},
            title: '',
            url: url,
        });
    };
    return {
        location: location,
        history: history,
        setHref: setHref,
    };
}
